### DISCUSSION TITLE
Removing duplicates from ff3 due to recent bug

### DISCUSSION THREAD

**Discussion by cruzzoe** [firefly-iii/firefly-iii#9238] [Firefly III]
TLDR: is it safe to delete dupes directly from table: transactions in the db? 

Due to the bug in 6.1.9 [link](https://github.com/firefly-iii/firefly-iii/issues/9155 ) I now have hundreds of duplicates for one of my accounts.

How did this happen? Every day I import using an automated job via goCardless. I create a custom internal ID via hashing and then use the ff3 importer docker img to load it in. The bug linked above prevented the duplicate checking from working inside ff3. This means each day, I have an additional transaction for EVERY transaction that is fed in from go cardless. This means I have hundreds to clean out from the db.

I dont want to mess anything up in the db, but cleaning from the UI is too time consuming. Is there anything I can do to speed this up? I was thinking of connecting to the db (MYSQL) and then using sql to tidy up the duplicate transactions. Please let me know if this reasonable or if I will break any of the other index/table joins etc...

Example

Two transactions created on consecutive days with the same details.
The internal hash id is the same
so....Firefly3 should see this as duplicates.
As it wasnt able to do this, I've now got 100s of this transaction. How do I resolve without many clicks in the UI>

Raw json1`
{"id":3664,"created_at":"2024-09-11T00:00:15.000000Z","updated_at":"2024-09-11T00:00:15.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"title":null,"transaction_journals":[{"id":3668,"created_at":"2024-09-11T00:00:15.000000Z","updated_at":"2024-09-11T00:00:15.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"transaction_type_id":3,"transaction_group_id":3664,"bill_id":null,"transaction_currency_id":3,"description":"my credit card","date":"2024-08-14T22:00:00.000000Z","interest_date":null,"book_date":null,"process_date":null,"order":0,"tag_count":0,"completed":true,"transactions":[{"id":7335,"created_at":"2024-09-11T00:00:15.000000Z","updated_at":"2024-09-11T00:00:15.000000Z","deleted_at":null,"reconciled":false,"account_id":318,"transaction_journal_id":3668,"description":null,"transaction_currency_id":3,"amount":"-68.150000000000","foreign_amount":"","foreign_currency_id":null,"identifier":0,"account":{"id":318,"created_at":"2023-11-26T01:02:07.000000Z","updated_at":"2024-08-16T10:40:54.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"account_type_id":3,"name":"First Direct","virtual_balance":"0.000000000000","iban":"","active":true,"order":6},"budgets":[],"categories":[]},{"id":7336,"created_at":"2024-09-11T00:00:15.000000Z","updated_at":"2024-09-11T00:00:15.000000Z","deleted_at":null,"reconciled":false,"account_id":341,"transaction_journal_id":3668,"description":null,"transaction_currency_id":3,"amount":"68.150000000000","foreign_amount":"","foreign_currency_id":null,"identifier":0,"account":{"id":341,"created_at":"2023-11-26T02:34:38.000000Z","updated_at":"2024-08-16T10:40:54.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"account_type_id":3,"name":"my card","virtual_balance":"0.000000000000","iban":"","active":true,"order":7},"budgets":[],"categories":[]}],"meta":[{"id":9824,"created_at":"2024-09-11T00:00:15.000000Z","updated_at":"2024-09-11T00:00:15.000000Z","transaction_journal_id":3668,"name":"internal_reference","data":"00952deff15e15cb058fcae866faa109","hash":"8a48155243e1273ac3a14c378962f5c8722d75dea0b7565cff2adb78abec334c","deleted_at":null},{"id":9825,"created_at":"2024-09-11T00:00:15.000000Z","updated_at":"2024-09-11T00:00:15.000000Z","transaction_journal_id":3668,"name":"import_hash_v2","data":"1b22d853d9025b0d5dc48d3c7f50b7ec0603ee47be9e300199a4878f74ed8af6","hash":"3dbb9688acb0021d6dc75a545d6209281e4cbf5fadc0104442c7c40d202ad479","deleted_at":null},{"id":9826,"created_at":"2024-09-11T00:00:15.000000Z","updated_at":"2024-09-11T00:00:15.000000Z","transaction_journal_id":3668,"name":"original_source","data":"ff3-v6.1.19|api-v2.1.0","hash":"bbefb1b6b10261fbe1dfd32a3d4ec8775274e206adbc70f5a54f91e824c7da78","deleted_at":null}],"tags":[{"id":287,"created_at":"2024-09-11T00:00:11.000000Z","updated_at":"2024-09-11T00:00:11.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"tag":"2024-09-11 @ 02:00 FD","tagMode":"nothing","date":"2024-09-10T22:00:00.000000Z","description":"","pivot":{"transaction_journal_id":3668,"tag_id":287}}],"categories":[],"budgets":[],"notes":[],"locations":[],"attachments":[],"links":[],"piggy_bank_events":[]}]}`

Raw json2
`{"id":3608,"created_at":"2024-09-10T00:00:16.000000Z","updated_at":"2024-09-10T00:00:16.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"title":null,"transaction_journals":[{"id":3612,"created_at":"2024-09-10T00:00:16.000000Z","updated_at":"2024-09-10T00:00:16.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"transaction_type_id":3,"transaction_group_id":3608,"bill_id":null,"transaction_currency_id":3,"description":"my credit card","date":"2024-08-14T22:00:00.000000Z","interest_date":null,"book_date":null,"process_date":null,"order":0,"tag_count":0,"completed":true,"transactions":[{"id":7223,"created_at":"2024-09-10T00:00:16.000000Z","updated_at":"2024-09-10T00:00:16.000000Z","deleted_at":null,"reconciled":false,"account_id":318,"transaction_journal_id":3612,"description":null,"transaction_currency_id":3,"amount":"-68.150000000000","foreign_amount":"","foreign_currency_id":null,"identifier":0,"account":{"id":318,"created_at":"2023-11-26T01:02:07.000000Z","updated_at":"2024-08-16T10:40:54.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"account_type_id":3,"name":"First Direct","virtual_balance":"0.000000000000","iban":"","active":true,"order":6},"budgets":[],"categories":[]},{"id":7224,"created_at":"2024-09-10T00:00:16.000000Z","updated_at":"2024-09-10T00:00:16.000000Z","deleted_at":null,"reconciled":false,"account_id":341,"transaction_journal_id":3612,"description":null,"transaction_currency_id":3,"amount":"68.150000000000","foreign_amount":"","foreign_currency_id":null,"identifier":0,"account":{"id":341,"created_at":"2023-11-26T02:34:38.000000Z","updated_at":"2024-08-16T10:40:54.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"account_type_id":3,"name":"my Clarity","virtual_balance":"0.000000000000","iban":"","active":true,"order":7},"budgets":[],"categories":[]}],"meta":[{"id":9656,"created_at":"2024-09-10T00:00:16.000000Z","updated_at":"2024-09-10T00:00:16.000000Z","transaction_journal_id":3612,"name":"internal_reference","data":"00952deff15e15cb058fcae866faa109","hash":"8a48155243e1273ac3a14c378962f5c8722d75dea0b7565cff2adb78abec334c","deleted_at":null},{"id":9657,"created_at":"2024-09-10T00:00:16.000000Z","updated_at":"2024-09-10T00:00:16.000000Z","transaction_journal_id":3612,"name":"import_hash_v2","data":"1b22d853d9025b0d5dc48d3c7f50b7ec0603ee47be9e300199a4878f74ed8af6","hash":"3dbb9688acb0021d6dc75a545d6209281e4cbf5fadc0104442c7c40d202ad479","deleted_at":null},{"id":9658,"created_at":"2024-09-10T00:00:16.000000Z","updated_at":"2024-09-10T00:00:16.000000Z","transaction_journal_id":3612,"name":"original_source","data":"ff3-v6.1.19|api-v2.1.0","hash":"bbefb1b6b10261fbe1dfd32a3d4ec8775274e206adbc70f5a54f91e824c7da78","deleted_at":null}],"tags":[{"id":286,"created_at":"2024-09-10T00:00:12.000000Z","updated_at":"2024-09-10T00:00:12.000000Z","deleted_at":null,"user_id":1,"user_group_id":1,"tag":"2024-09-10 @ 02:00 FD","tagMode":"nothing","date":"2024-09-09T22:00:00.000000Z","description":"","pivot":{"transaction_journal_id":3612,"tag_id":286}}],"categories":[],"budgets":[],"notes":[],"locations":[],"attachments":[],"links":[],"piggy_bank_events":[]}]}`

**Comment by JC5** [comment]
Sorry, i completely forgot about your ticket. It arrived during a week abroad and I lost track of it.

What works is when you track the duplicate externalIDs in the table journal_meta, and then remove the associated entries (based on transaction_journal_id) from the table "transaction_journals". 

Rebooting the image should clean up any associated tables. Of course, make sure you have a backup.

**Comment by cruzzoe** [comment]
Hey James, After playing around a bit I managed to get it to work! Its better than deleting transactions but still quite length and Im concerned about trying to automate it fully incase I go too far....

Out of interest, why does the rest of the ff3 users not get impacted by this issue? Do they not rely on the internal_id for dupe checking with gocardless importer?

**Reply by JC5** [reply]
> Out of interest, why does the rest of the ff3 users not get impacted by this issue? Do they not rely on the internal_id for dupe checking with gocardless importer?

Honestly I'm not sure. I would expect my own installation to be impacted as well. But also, perhaps 6.1.20 (out now!) won't fix this issue for your specifically.

**Reply by cruzzoe** [reply]
I can't upgrade until the issue is resolved as every time I do a daily go careless pull, if I get 30 days of history then on a daily basis I end up with an additional 30days worth of duplicates!

When cleaning up today I found thousands of entries to remove.

I assume I'm hitting the issue as I rely on the CMD line to upload my transactions. 

The code to upload is:

            "docker", "run",
            "--rm",
            "-v", f"{self.import_dir}:/import",
            "-e", f"FIREFLY_III_ACCESS_TOKEN={TOKEN}",
            "-e", "IMPORT_DIR_ALLOWLIST=/import",
            "-e", f"FIREFLY_III_URL={HOME_IP}:8995",
            "-e", "WEB_SERVER=false",
            "fireflyiii/data-importer:latest"

And the importer config is:

    "duplicate_detection_method": "cell",
    "ignore_duplicate_lines": true,
    "unique_column_index": 8,
    "unique_column_type": "internal_reference",

Would you be able to give me any pointers on which bit of the code to look at? Will try to control fix if I can although I'm not familiar with php.

**Reply by JC5** [reply]
But the issue should be solved with 6.1.20, for new transactions for sure. I would suggest to import less transactions, especially when you're running a daily import.

**Comment by cruzzoe** [comment]
Oh I misunderstood. id thought you were saying it's not resolved in the new
version. I can limit the imported transactions to just the run date. That
might help...
